frame/staking/src/slashing.rs

```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
+		// as seed, it's better than a static value.
+		entropy.as_mut().copy_from_slice(parent_hash.as_ref());
+	}
+	entropy
+}
+
+// Decides if a validator should be disabled or not based on its offence. The bigger the offence -
+// the higher chance to disable the validator in question.
+fn disable_rand<T: Config>(offence: &Perbill) -> bool {
+	let entropy = compute_entropy::<T>(<frame_system::Pallet<T>>::parent_hash());
+	let mut rng = rand_chacha::ChaChaRng::from_seed(entropy.into());
+	let r = rng.next_u32() % 100;
+
+	Perbill::from_percent(r) <= *offence
+}
+
+// Gets a Vec and its desired length as an input. Randomly removes values from the Vec until its
+// length matches the desired length. The validator length should be bigger than the desired one.
+fn remove_random_validators<T: Config>(validators: Vec<u32>, desired_len: usize) -> Vec<u32> {
+	debug_assert!(validators.len() > desired_len);
+
+	let entropy = compute_entropy::<T>(<frame_system::Pallet<T>>::parent_hash());
+	let mut rng = rand_chacha::ChaChaRng::from_seed(entropy.into());
+
+	let mut indecies_to_remove = BTreeSet::new();
+	for _ in 0..=desired_len.saturating_sub(validators.len()) {
+		indecies_to_remove.insert(rng.next_u32() as usize % validators.len());
+	}
+
+	validators
+		.into_iter()
+		.enumerate()
+		.filter(|(pos, _)| !indecies_to_remove.contains(pos))
+		.map(|(_, val)| val)
+		.collect::<Vec<_>>()
+}
+
+/// Disable the validator with the id in `new_offender`. There are two possible cases:
+/// 1. The total number of disabled validators is below the threshold. Then disable the new one and
+///    finish.
+/// 2. The total number of disabled validators is equal or above the threshold. In this case add the
+///    disabled validators list is reshuffled. The reshuffling works by disabling the validator with
+///    a probability equal to its offence. If the disabled validators list ends up bigger than the
+///    threshold - validators are randomly removed from the list until the desired length is
+///    achieved.
+/// In both cases Session is notified for the change via the `SessionInterface.
+///
+/// NOTE: in case 2 the new offender might not be disabled.
+fn disable_new_offender<T: Config>(current_offenders: &Vec<(u32, Perbill)>, new_offender: u32) {
+	const LIMIT: usize = 42; // TODO: extract this as a parameter
+
+	let currently_disabled = DisabledOffenders::<T>::get();
+
+	if currently_disabled.len() < LIMIT {
+		// we are below the limit - just disable
+		T::SessionInterface::disable_validator(new_offender);
+		add_to_disabled_offenders::<T>(new_offender);
+
+		return
+	}
+
+	// Now selectively disable based on the offence
+	let mut disabled = Vec::new();
+	for (v, o) in current_offenders {
+		if disable_rand::<T>(o) {
+			disabled.push(*v);
+		}
+	}
+
+	let disabled = if disabled.len() > LIMIT {
+		// remove `disabled.len() - LIMIT` random elements from `disabled`
+		remove_random_validators::<T>(disabled, LIMIT)
+	} else {
+		disabled
+	};
```

Comment:
What about the new offender?

Reply:
Yes, this is one of the shortcuts I've made and not sure if it is sound or not.

The new offender might not get disabled thanks to the reshuffling. This sounds like a bad idea - I'll fix it.